clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java

```diff
@@ -0,0 +1,260 @@
+    }
+
+    boolean notInGroup() {
+        return membershipManager.state() == MemberState.UNJOINED ||
+                membershipManager.state() == MemberState.FAILED;
+    }
+
+    private NetworkClientDelegate.UnsentRequest makeHeartbeatRequest() {
+        ConsumerGroupHeartbeatRequestData data = new ConsumerGroupHeartbeatRequestData()
+            .setGroupId(membershipManager.groupId())
+            .setMemberEpoch(membershipManager.memberEpoch())
+            .setMemberId(membershipManager.memberId())
+            .setRebalanceTimeoutMs(rebalanceTimeoutMs);
+
+        membershipManager.groupInstanceId().ifPresent(data::setInstanceId);
+
+        if (this.subscriptions.hasPatternSubscription()) {
+            // We haven't discsussed how Regex is stored in the consumer. We could do it in the subscriptionState
+            // , in the memberStateManager, or here.
+            // data.setSubscribedTopicRegex(regex)
+        } else {
+            data.setSubscribedTopicNames(new ArrayList<>(this.subscriptions.subscription()));
+        }
+
+        // stubbed
+        data.setServerAssignor("stubbed.assignor");
+
+        NetworkClientDelegate.UnsentRequest request = new NetworkClientDelegate.UnsentRequest(
+            new ConsumerGroupHeartbeatRequest.Builder(data),
+            coordinatorRequestManager.coordinator());
+
+        request.future().whenComplete((response, exception) -> {
+            final long currentTimeMs = time.milliseconds();
+            if (exception == null) {
+                onSuccess((ConsumerGroupHeartbeatResponse) response.responseBody(), currentTimeMs);
+            } else {
+                onFailure(exception, currentTimeMs);
+            }
+        });
+        return request;
+    }
+
+    private void onFailure(final Throwable exception, final long responseTimeMs) {
+        this.heartbeatRequestState.onFailedAttempt(responseTimeMs);
+        logger.debug("failed sending heartbeat due to {}", exception.getMessage());
+    }
+
+    private void onSuccess(final ConsumerGroupHeartbeatResponse response, long currentTimeMs) {
+        if (response.data().errorCode() == Errors.NONE.code()) {
+            this.heartbeatRequestState.onSuccessfulAttempt(currentTimeMs);
+            this.heartbeatRequestState.reset();
+            try {
+                membershipManager.updateState(response.data());
+            } catch (KafkaException e) {
+                logger.error("Received unexpected error in heartbeat response: {}", e.getMessage());
+            }
+            return;
+        }
+
+        onError(response, currentTimeMs);
+    }
+
+    private void onError(final ConsumerGroupHeartbeatResponse response,
+                         final long currentTimeMs) {
+
+        this.heartbeatRequestState.onFailedAttempt(currentTimeMs);
+        short errorCode = response.data().errorCode();
+        if (errorCode == Errors.NOT_COORDINATOR.code() || errorCode == Errors.COORDINATOR_NOT_AVAILABLE.code()) {
```

Comment:
It might be helpful to read the test: `public void testHeartbeatResponse_errorHandling(final Errors error, final boolean isFatal)`